### PR TITLE
Add SyncFusedTensorOp to fix fuse_optimizer_op_pass bug

### PR DESCRIPTION
**Pull request by chenwhql**
In the GPU training environment, if the `fuse_all_optimizer_ops` strategy is opened, ParallelExecutor will fused the input gradients of optimizer ops to a contiguous address on the GPU device **before the main graph is executed**, since some grad operators may not have a GPU kernel, the gradients gerarated cannot be written to the GPU device as expected, but generated separately on the CPU device, which will cause the program to use dirty data in GPU device to continue training, it may cause calculate error or NAN. 

Related PR：https://github.com/PaddlePaddle/Paddle/pull/19966

This PR insert a operator **SyncFusedTensorOp** into graph to solve this problem. When this problem happened, this operator will synchronize the output on the CPU device to the GPU device to ensure the correctness of data.

I construct a test case by deleting the CUDA kernel of mul_op, so in a simple fc net, the input of optimizer op grad fc_0.w_0@GRAD  should be copy form CPUPlace to CUDAPlace, the test log is as follows.

```
I1017 06:23:17.341679  7481 fused_all_reduce_op_handle.cc:48] fused_all_reduce(name:fc_0.b_0@GRAD, place:CUDAPlace(0), version:0, scope_idx:0, name:fc_0.b_0@GRAD, place:CUDAPlace(1), version:0, scope_idx:1, name:fc_0.w_0@GRAD, place:CUDAPlace(0), version:0, scope_idx:0, name:fc_0.w_0@GRAD, place:CUDAPlace(1), version:0, scope_idx:1, ) --> (name:fc_0.b_0@GRAD, place:CUDAPlace(0), version:1, scope_idx:0, name:fc_0.b_0@GRAD, place:CUDAPlace(1), version:1, scope_idx:1, name:fc_0.w_0@GRAD, place:CUDAPlace(0), version:1, scope_idx:0, name:fc_0.w_0@GRAD, place:CUDAPlace(1), version:1, scope_idx:1, )
I1017 06:23:17.341895  7481 operator.cc:165] CUDAPlace(0) Op(sync_fused_tensor), inputs:{X[fc_0.b_0@GRAD:float[1]({}), fc_0.w_0@GRAD:float[13, 1]({})]}, outputs:{Out[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})]}.
I1017 06:23:17.342021  7481 operator.cc:988] expected_kernel_key:data_type[float]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN]
I1017 06:23:17.342046  7481 operator.cc:1086] Transform Variable fc_0.w_0@GRAD from data_type[float]:data_layout[NCHW]:place[CPUPlace]:library_type[PLAIN] to data_type[float]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN]
I1017 06:23:17.342068  7481 scope.cc:164] Create variable fc_0.w_0@GRAD
I1017 06:23:17.342087  7481 data_device_transform.cc:21] DeviceTransform in, src_place CPUPlace dst_place: CUDAPlace(0)
I1017 06:23:17.342111  7481 tensor_util.cu:122] TensorCopySync 13, 1 from CPUPlace to CUDAPlace(0)
I1017 06:23:17.342185  7481 operator.cc:185] CUDAPlace(0) Op(sync_fused_tensor), inputs:{X[fc_0.b_0@GRAD:float[1]({}), fc_0.w_0@GRAD:float[13, 1]({})]}, outputs:{Out[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})]}.
I1017 06:23:17.342227  7481 operator.cc:165] CUDAPlace(0) Op(momentum), inputs:{Grad[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})], LearningRate[learning_rate_0:float[1]({})], Param[@FUSEDVAR@_momentum_Param_fc_0.b_0:float[128]({})], Velocity[@FUSEDVAR@_momentum_Velocity_fc_0.b_0_velocity_0:float[128]({})]}, outputs:{ParamOut[@FUSEDVAR@_momentum_Param_fc_0.b_0:float[128]({})], VelocityOut[@FUSEDVAR@_momentum_Velocity_fc_0.b_0_velocity_0:float[128]({})]}.
```

### CODE REVIEW

**Comment by Xreki** [comment]
我有些疑问：
- coalese_tensor是用来准备optimizer的输入数据的，将多个grad拷贝到一段连续的memory里面。如果多个Input，有的在CPU上，有的在GPU上，该op使用默认的`GetExpectedKernelType`，也就是会选择GPU kernel，正常情况下，会产生data transform，自动将CPU数据拷贝到GPU上。确定需要手动拷贝？你log里面的一些，就是在做data transform
```
I1017 06:23:17.341679  7481 fused_all_reduce_op_handle.cc:48] fused_all_reduce(name:fc_0.b_0@GRAD, place:CUDAPlace(0), version:0, scope_idx:0, name:fc_0.b_0@GRAD, place:CUDAPlace(1), version:0, scope_idx:1, name:fc_0.w_0@GRAD, place:CUDAPlace(0), version:0, scope_idx:0, name:fc_0.w_0@GRAD, place:CUDAPlace(1), version:0, scope_idx:1, ) --> (name:fc_0.b_0@GRAD, place:CUDAPlace(0), version:1, scope_idx:0, name:fc_0.b_0@GRAD, place:CUDAPlace(1), version:1, scope_idx:1, name:fc_0.w_0@GRAD, place:CUDAPlace(0), version:1, scope_idx:0, name:fc_0.w_0@GRAD, place:CUDAPlace(1), version:1, scope_idx:1, )
I1017 06:23:17.341895  7481 operator.cc:165] CUDAPlace(0) Op(sync_fused_tensor), inputs:{X[fc_0.b_0@GRAD:float[1]({}), fc_0.w_0@GRAD:float[13, 1]({})]}, outputs:{Out[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})]}.
I1017 06:23:17.342021  7481 operator.cc:988] expected_kernel_key:data_type[float]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN]
I1017 06:23:17.342046  7481 operator.cc:1086] Transform Variable fc_0.w_0@GRAD from data_type[float]:data_layout[NCHW]:place[CPUPlace]:library_type[PLAIN] to data_type[float]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN]
I1017 06:23:17.342068  7481 scope.cc:164] Create variable fc_0.w_0@GRAD
I1017 06:23:17.342087  7481 data_device_transform.cc:21] DeviceTransform in, src_place CPUPlace dst_place: CUDAPlace(0)
I1017 06:23:17.342111  7481 tensor_util.cu:122] TensorCopySync 13, 1 from CPUPlace to CUDAPlace(0)
I1017 06:23:17.342185  7481 operator.cc:185] CUDAPlace(0) Op(sync_fused_tensor), inputs:{X[fc_0.b_0@GRAD:float[1]({}), fc_0.w_0@GRAD:float[13, 1]({})]}, outputs:{Out[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})]}.
```

**Comment by chenwhql** [comment]
> 我有些疑问：
> 
> * coalese_tensor是用来准备optimizer的输入数据的，将多个grad拷贝到一段连续的memory里面。如果多个Input，有的在CPU上，有的在GPU上，该op使用默认的`GetExpectedKernelType`，也就是会选择GPU kernel，正常情况下，会产生data transform，自动将CPU数据拷贝到GPU上。确定需要手动拷贝？你log里面的一些，就是在做data transform
> 
> ```
> I1017 06:23:17.341679  7481 fused_all_reduce_op_handle.cc:48] fused_all_reduce(name:fc_0.b_0@GRAD, place:CUDAPlace(0), version:0, scope_idx:0, name:fc_0.b_0@GRAD, place:CUDAPlace(1), version:0, scope_idx:1, name:fc_0.w_0@GRAD, place:CUDAPlace(0), version:0, scope_idx:0, name:fc_0.w_0@GRAD, place:CUDAPlace(1), version:0, scope_idx:1, ) --> (name:fc_0.b_0@GRAD, place:CUDAPlace(0), version:1, scope_idx:0, name:fc_0.b_0@GRAD, place:CUDAPlace(1), version:1, scope_idx:1, name:fc_0.w_0@GRAD, place:CUDAPlace(0), version:1, scope_idx:0, name:fc_0.w_0@GRAD, place:CUDAPlace(1), version:1, scope_idx:1, )
> I1017 06:23:17.341895  7481 operator.cc:165] CUDAPlace(0) Op(sync_fused_tensor), inputs:{X[fc_0.b_0@GRAD:float[1]({}), fc_0.w_0@GRAD:float[13, 1]({})]}, outputs:{Out[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})]}.
> I1017 06:23:17.342021  7481 operator.cc:988] expected_kernel_key:data_type[float]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN]
> I1017 06:23:17.342046  7481 operator.cc:1086] Transform Variable fc_0.w_0@GRAD from data_type[float]:data_layout[NCHW]:place[CPUPlace]:library_type[PLAIN] to data_type[float]:data_layout[ANY_LAYOUT]:place[CUDAPlace(0)]:library_type[PLAIN]
> I1017 06:23:17.342068  7481 scope.cc:164] Create variable fc_0.w_0@GRAD
> I1017 06:23:17.342087  7481 data_device_transform.cc:21] DeviceTransform in, src_place CPUPlace dst_place: CUDAPlace(0)
> I1017 06:23:17.342111  7481 tensor_util.cu:122] TensorCopySync 13, 1 from CPUPlace to CUDAPlace(0)
> I1017 06:23:17.342185  7481 operator.cc:185] CUDAPlace(0) Op(sync_fused_tensor), inputs:{X[fc_0.b_0@GRAD:float[1]({}), fc_0.w_0@GRAD:float[13, 1]({})]}, outputs:{Out[@FUSEDVAR@@GRAD@fc_0.b_0@GRAD:float[128]({})]}.
> ```

这个op是先于整体graph执行的，在PE初始化变量的时候，根据param的维度申请了grad的GPU内存，运行时就不会再拷贝了，这个op在一个独立的临时program里面运行的

https://github.com/PaddlePaddle/Paddle/blob/a9bc92c31422674eab2ce4b7ddc2282f9a727499/paddle/fluid/framework/details/scope_buffered_ssa_graph_executor.cc#L117

File: paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc [hunk 1]
```diff
@@ -336,10 +345,71 @@ proto::VarType::Type FuseOptimizerOpPass::GetTypeOfVar(
...
+    ir::Node *fused_opt_node, ir::Graph *result) const {
+  for (auto in_var_node : in_var_nodes) {
+    // set outpus
+    ir::Node *dep_var = result->CreateControlDepVar();
+    VLOG(6) << "Insert dep_var between " << sync_tensor_op_node->Op()->Type()
+            << " and " << fused_opt_node->Op()->Type();
+    sync_tensor_op_node->outputs.emplace_back(dep_var);
+    fused_opt_node->inputs.emplace_back(dep_var);
+    dep_var->inputs.emplace_back(sync_tensor_op_node);
+    dep_var->outputs.emplace_back(fused_opt_node);
```
Comment (chengduoZH) [inline paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc:389]: It seems that there not need so many `dep_var` between `sync_tensor_op_node ` and `fused_opt_node `.

File: paddle/fluid/operators/sync_fused_tensor_op.cu [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace ops = paddle::operators;
+namespace plat = paddle::platform;
+REGISTER_OP_CUDA_KERNEL(
+    sync_fused_tensor,
+    ops::SyncFusedTensorOpKernel<paddle::platform::CUDADeviceContext,
+                                 plat::float16>,
+    ops::SyncFusedTensorOpKernel<paddle::platform::CUDADeviceContext, int>,
+    ops::SyncFusedTensorOpKernel<paddle::platform::CUDADeviceContext, float>,
+    ops::SyncFusedTensorOpKernel<paddle::platform::CUDADeviceContext, double>);
```
Comment (chengduoZH) [inline paddle/fluid/operators/sync_fused_tensor_op.cu:26]: In fact that, you don't need to define OpKernel here, because the logic of this operator are the same.

**Comment by Xreki** [comment]
> 这个op是先于整体graph执行的，在PE初始化变量的时候，根据param的维度申请了grad的GPU内存，运行时就不会再拷贝了，这个op在一个独立的临时program里面运行的

明白了。

在此记录一下我对这个工作的学习：

- fused_optimizer_ops的原理，是将多个需要optimizer更新的Tensor拼凑成一块连续的空间，这样就只需要调用一次optimizer op就能将所有的Tensor一次更新完。为了实现这个功能，引入了`coalesce_tensor`这样一个op。这个op的主要功能是：**假设有N个Tensor要通过optimizer来更新，但他们的地址不是连续的，coalesce_tensor op中申请了一大段连续的空间，先将这N个Tensor的数据拷贝到这个连续的内存/显存空间，然后让这N个Tensor指向这些地址，相当于对这N个Tensor的内存进行了一次重新分配。因此，coalesce_tensor只需在训练开始前执行一次。** 具体参数介绍如下：

### coalesce_tensor
- Inputs
    - Input：`std::vector<Tensor>`，Tensor向量。每个输入Tensor可以是多维的，并且维度可以各不相同，但要求dtype必须相同。
- Outputs
    - FusedOutput：1-D Tensor，其长度是Input中各个Tensor元素个数之和。**注意：FusedOutput会调用mutable_data申请空间。**
    - Output：`std::vector<Tensor>`，LoDTensor向量。其中Output的LoDTensor个数必须和Input的LoDTensor个数相同。是FusedOutput的一部分。和Input一一对应，和FusedOutput共享内存。**注意：当Input和Output的name都一样时，其实相当于对Input Tensor的内存进行了重新分配，要求Input Tensor的内存不能被GC释放且dims不能改变。**
- Attrs:
    - check_name：如果设置成True，则会检查Input和Output的名字，要求Input和Output的名字必须完全一样。如果设置成False，则会将Output resize成和Input一样的dims。
    - copy_data：如果设置成True，则会将Input中各个Tensor的元素拷贝到FusedOutput中。每个Tensor拷贝后都会做对齐处理。
    - set_constant：当copy_data设置成False，set_contant设置成True时，会将整个FusedOutput设置成一个常量值。**注意：copy_data和set_contant都为False时，FusedOutput是没有初始化的。**

File: paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/framework/ir/fuse_optimizer_ops_pass/fuse_optimizer_op_pass.cc:389]: change to close this strategy in this scenario, new PR: https://github.com/PaddlePaddle/Paddle/pull/21005

File: paddle/fluid/operators/sync_fused_tensor_op.cu (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/operators/sync_fused_tensor_op.cu:26]: change to close this strategy in this scenario, new PR: https://github.com/PaddlePaddle/Paddle/pull/21005
